markdown/english/3-JobPhysicalPlan.md

```diff
@@ -0,0 +1,247 @@
+
+## A Complex Logical Plan
+
+![ComplexJob](../PNGfigures/ComplexJob.png)
+The code of this application is attached at the end of this chapter.
+
+**How to properly define stages and determine the tasks with such a complex data dependency graph?**
+
+An intuitive idea is to associate one RDD and its preceding RDD to form a stage, in this way each arrow in the above graph will become a task. For the case of 2 RDDs aggregates into one, we may create a stage with these 3 RDDs. This strategy could be a working solution, but will not be efficient. It has a subtle, but severe problem: **lots of intermediate data needs to be stored**. For a physical task, its result will be stored either on local disk, or in the memory, or both. If a task is generated for each arrow in the data dependency graph, the system needs to store data of all the RDDs. It will cost a lot.
+
+If we examine the logical plan more closely, we may find out that in each RDD, the partitions are independent from each other. That is to say, inside each RDD, the data within a partition will not interfere others. With this observation, an aggressive idea is to consider the whole diagram as a single stage and create one physical task for each partition of the final RDD (`FlatMappedValuesRDD`). The following diagram illustrates this idea:
+
+![ComplexTask](../PNGfigures/ComplexTask.png)
+
+All thick arrows in above diagram belong to task1 whose result is the first partition of the final RDD of the job. Note that in order to compute the first partition of the `CoGroupedRDD`, we need to evaluate all partitions of its preceding RDDs since it's a `ShuffleDependency`. So in the computation of task1, we can also take the chance to compute the `CoGroupedRDD`'s second and third partition, for task2 and task3 respecively. As a result, the task2 and task3 are simpler. They are represented by thin arrows and dashed arrows in the diagram.
+
+However, there's 2 problems with this idea:
+  - The first task is big. Because of the `ShuffleDependency`, we have to evaluate all the partitions of the preceding RDDs.
+  - Need to design clever algorithms to determine which are the partitions to cache.
+
+But there's also one good point in this idea, that is to **pipeline the data: the data is computed when they are actually needed in a flow fashion**. For example in the first task, we check backwards from the final RDD (`FlatMappedValuesRDD`) to see which are the RDDs and partitions that are actually needed to be evaluated. And between the RDDs with `NarrowDependency` relation, no intermediate data needs to be stored.
+
+It will be clearer to understand the pipelining if we consider a record-level point of view. The following diagram illustrates different evaluation patterns for RDDs with `NarrowDependency`.
+
+![Dependency](../PNGfigures/pipeline.png)
+
+The first pattern (pipeline pattern) is equivalent to:
+
+```scala
```

Comment:
I add these 2 blocks of pseudo-code to illustrate the pipeline style of processing records.